File: 프로그래머스/2/86052. 빛의 경로 사이클/빛의 경로 사이클.java
import java.util.*;
class Solution {
    static int[][] directions = {{1,0},{0,1},{-1,0},{0,-1}};
    class Light {
        int r;
        int c;
        int direction;
        public Light(int r, int c, int direction){
            this.r = r;
            this.c = c;
            this.direction = direction;
        }
    }
    public int[] solution(String[] grid) {
        char[][] map = new char[grid.length][grid[0].length()];
        for(int r=0; r<grid.length; r++){
            for(int c=0; c<grid[0].length(); c++){
                map[r][c] = grid[r].charAt(c);
            }
        }
        boolean[][][] visited = new boolean[map.length][map[0].length][4];
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for(int r=0; r<map.length; r++){
            for(int c=0; c<map[0].length; c++){
                for(int d=0; d<4; d++){
                    if(!visited[r][c][d]){
                        pq.offer(DFS(new Light(r, c, d), map, visited));
                    }
                }
            }
        }
        int[] answer = new int[pq.size()];
        int index = 0;
        while(!pq.isEmpty()){
            answer[index] = pq.poll();
            index++;
        }
        return answer;
    }
    public int DFS(Light light, char[][] map, boolean[][][] visited){
        int depth = 1;
        int r = light.r;
        int c = light.c;
        int direction = light.direction;
        visited[r][c][direction] = true;
        while(true){
            int nr = (r + directions[direction][0] + map.length) % map.length;
            int nc = (c + directions[direction][1] + map[0].length) % map[0].length;
            int nextDirection = 0;
            if(map[nr][nc] == 'S'){
                nextDirection = direction;
            } else if(map[nr][nc] == 'L'){
                nextDirection = (direction + 1) % 4;
            } else {
                nextDirection = (direction + 3) % 4;
            }
            if(!visited[nr][nc][nextDirection]){
                visited[nr][nc][nextDirection] = true;
                depth++;
                r = nr;
                c = nc;
                direction = nextDirection;
            } else {
                break;
            }
        }
        return depth;
    }
}

File: 프로그래머스/2/86052. 빛의 경로 사이클/빛의 경로 사이클.py
def solution(grid):
    directions = [[1,0],[0,1],[-1,0],[0,-1]]
    rLen = len(grid)
    cLen = len(grid[0])
    gridMap = [[""] * cLen for r in range(rLen)]
    for r in range(rLen):
        for c in range(cLen):
            gridMap[r][c] = grid[r][c]
    visited = [[[False] * 4 for _ in range(cLen)] for _ in range(rLen)]
    
    class Light:
        def __init__(self, r, c, direction):
            self.r = r
            self.c = c
            self.direction = direction
    
    def DFS(light, gridMap, visited):
        r = light.r
        c = light.c
        direction = light.direction
        visited[r][c][direction] = True
        depth = 1
        while True:
            nr = (r + directions[direction][0] + rLen) % rLen
            nc = (c + directions[direction][1] + cLen) % cLen
            if gridMap[nr][nc] == "S":
                nextDirection = direction
            elif gridMap[nr][nc] == "L":
                nextDirection = (direction + 1) % 4
            else:
                nextDirection = (direction + 3) % 4
            
            if not visited[nr][nc][nextDirection]:
                visited[nr][nc][nextDirection] = True
                depth += 1
                r = nr
                c = nc
                direction = nextDirection
            else:
                break
        
        return depth
    
    answer = []
    for r in range(rLen):
        for c in range(cLen):
            for d in range(4):
                if not visited[r][c][d]:
                    light = Light(r, c, d)
                    answer.append(DFS(light, gridMap, visited))
                    
    answer.sort()
    return answer